### PR TITLE
Add MapParam DSL for HTTP

### LINKED ISSUE #1479: Properly error out when an unsupported type is used as query string
`goa` generates incorrect code instead of erroring out when using a query string parameter of type `MapOf(String, Any)` for example. Properly catch unsupported types and report an error instead. Same applies to headers and path parameters.

### PR DESCRIPTION
**Pull request by nitinmohan87**
Fixes #1479 
Adds a MapParams DSL to the HTTP endpoint expression which maps the query params to an entire payload (of map type) or an attribute in the payload object (attribute must be map type).

### CODE REVIEW

File: http/codegen/server.go [hunk 1]
```diff
@@ -443,6 +444,27 @@ func {{ .RequestDecoder }}(mux goahttp.Muxer, decoder func(*http.Request) goahtt
...
+		{{- if .Required }}
+		if len({{ .VarName }}Raw) == 0 {
+			err = goa.MergeErrors(err, goa.MissingFieldError("{{ .Name }}", "query string"))
+		}
+		{{- end }}
+
+		{{- if not .Required }}
+		if len({{ .VarName }}Raw) != 0 {
+		{{- end }}
+			err2 := json.Unmarshal([]byte({{ .VarName }}Raw), &{{ .VarName }})
```
Comment (raphael) [inline http/codegen/server.go]: how does this work? {{ .VarName }}Raw would be of type `url.Values` which is `map[string][]string` - how can this be cases to `[]byte` and then JSON unmarhsalled? In general I don't think we should use JSON to marhsal/unmarshal. The encoding / decoding code everywhere type asserts or type converts as needed and we should do that here too.

File: http/design/endpoint.go [hunk 2]
```diff
@@ -36,6 +36,12 @@ type (
 		// Metadata is a set of key/value pairs with semantic that is
 		// specific to each generator.
 		Metadata design.MetadataExpr
+		// MapQueryParams indicates that the query params are mapped to the
+		// payload in the method expression. If the pointer refers
+		// to a non-empty string, the query params are mapped to an attribute
+		// in the payload with the same name. If the pointer empty string,
```
Comment (raphael) [inline http/design/endpoint.go]: missing "is an"

File: http/codegen/client.go [hunk 3]
```diff
@@ -204,7 +218,7 @@ func {{ .RequestEncoder }}(encoder func(*http.Request) goahttp.Encoder) func(*ht
 			{{- end }}
 		{{- end }}
 	{{- end }}
-	{{- if .Payload.Request.QueryParams }}
+	{{- if or .Payload.Request.QueryParams }}
```
[the remark this answers is not in the document]
Comment (nitinmohan87) [reply http/codegen/client.go]: My mistake. Will take it out.

[the remark this answers is not in the document]
Comment (raphael) [reply http/codegen/client.go]: cool you probably know already but it was in another place as well.

File: http/codegen/server.go (hunk 1 above)
Comment (nitinmohan87) [reply http/codegen/server.go]: `r.URL.Query().Get()` returns a string and it gets unmarshalled into `map[string][]string`.

**Comment by raphael** [comment]
Thank you!

File: http/codegen/client.go [hunk 4]
```diff
@@ -218,11 +218,24 @@ func {{ .RequestEncoder }}(encoder func(*http.Request) goahttp.Encoder) func(*ht
...
+		for key, value := range p{{ if .FieldName }}.{{ .FieldName }}{{ end }} {
+			{{ template "type_conversion" (typeConversionData .Type.KeyType.Type "keyStr" "key") }}
+			{{- if eq .Type.ElemType.Type.Name "array" }}
+			for _, val := range value {
+				{{ template "type_conversion" (typeConversionData .Type.ElemType.Type.ElemType.Type (printf "%sStr" "val") "val") }}
+				values.Add(keyStr, valStr)
+			}
+			{{- else }}
+			{{ template "type_conversion" (typeConversionData .Type.ElemType.Type (printf "%sStr" "value") "value") }}
+      values.Add(keyStr, valueStr)
```
Comment (raphael) [inline http/codegen/client.go]: indenting ^

**Pull request merged**